lib/cn/cn_tree.c

```diff
@@ -1443,22 +1483,168 @@ cn_comp_update_spill(struct cn_compaction_work *work, struct kvset **kvsets)
-    cn_comp_update_spill(work, kvsets);
-    return 0;
+    INIT_LIST_HEAD(&retired_kvsets);
+
+    /* Allocate a new left node and add the split output kvsets on the left to this node.
+     * This need not be done under the tree lock as this new node is not published yet.
+     */
+    if (nodeidv[0] != CN_TREE_INVALID_NODEID) {
+        left = cn_node_alloc(tree, nodeidv[0]);
+        if (!left)
+            return merr(ENOMEM);
+
+        for (k = 0; k < w->cw_kvset_cnt; k++) {
+            if (kvsets[k])
+                kvset_list_add_tail(kvsets[k], &left->tn_kvset_list);
+        }
+
+        w->cw_split.nodev[0] = left;
+    }
+
+    if (nodeidv[1] != CN_TREE_INVALID_NODEID) {
+        /* pnode is protected by an exclusive compaction token, so the max key cannot
+         * change while a node split is in progress
+         */
+        cn_tree_node_get_max_key(pnode, rekey, sizeof(rekey), &reklen);
+    }
+
+    rmlock_wlock(&tree->ct_lock);
+    {
+        /* Move all kvsets from parent node to the retired list.
+         */
+        for (k = 0; k < w->cw_kvset_cnt; k++) {
+            assert(!list_empty(&pnode->tn_kvset_list));
+            le = list_last_entry(&pnode->tn_kvset_list, struct kvset_list_entry, le_link);
+            list_del(&le->le_link);
+            list_add(&le->le_link, &retired_kvsets);
+        }
+        assert(list_empty(&pnode->tn_kvset_list));
+
+        /* Repurpose the parent node as the right node and add kvsets to it.
+         * The parent node is already part of the cN tree and is protected by a compaction
+         * token.
+         */
+        right = pnode;
+        if (nodeidv[1] != CN_TREE_INVALID_NODEID) {
+            right->tn_nodeid = nodeidv[1];
+
+            INIT_LIST_HEAD(&right->tn_kvset_list);
+            for (k = w->cw_kvset_cnt; k < w->cw_outc; k++) {
+                if (kvsets[k])
+                    kvset_list_add_tail(kvsets[k], &right->tn_kvset_list);
+            }
+
+            /* The last node in the route map contains all keys that are greater than the
+             * penultimate node. Under rare circumstances the split key chosen for the last
+             * node can be lexicographically greater than its edge key. The below logic
+             * detects this situation and updates the edge key of the right node to its
+             * max key at the time of split.
+             */
+            if (route_node_islast(right->tn_route_node)) {
+                int rc = route_node_keycmp(right->tn_route_node, w->cw_split.key, w->cw_split.klen);
+                if (ev(rc <= 0)) {
+                    err = route_node_key_modify(tree->ct_route_map, right->tn_route_node,
+                                                rekey, reklen);
+                    if (err) {
+                        rmlock_wunlock(&tree->ct_lock);
+                        goto exit;
+                    }
+                }
+            }
```

Comment:
Mabye we should set the edge key of the right most node to 0xffff...fff at kvs open time.  Then we wouldn't need special cases like this.  Root spill will also need this special case.

Reply:
Yes, that works. I added this special case as we are yet to finalize on the edge key for the rightmost node.

We will need to set a special edge key for the rightmost node during both kvs open and when opening an empty tree or a tree which is yet to have level-1 nodes.

Yes, we already special case root spill today using 'ri.last_node'.